presto-spark/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -0,0 +1,342 @@
+        }
+
+        @Override
+        public List<List<Object>> execute()
+        {
+            PlanFragment rootFragment = prestoSparkPlan.getPlan().getFragment();
+            RemoteSourceNode remoteSource = getOnlyElement(rootFragment.getRemoteSourceNodes());
+            PrestoSparkTaskDescriptor taskDescriptor = new PrestoSparkTaskDescriptor(
+                    session.toSessionRepresentation(),
+                    session.getIdentity().getExtraCredentials(),
+                    rootFragment,
+                    ImmutableList.of(),
+                    prestoSparkPlan.getTableWriteInfo());
+            SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(sparkTaskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
+
+            List<Tuple2<Integer, SerializedPrestoSparkPage>> resultRdd;
+            try {
+                List<Tuple2<Integer, SerializedPrestoSparkPage>> sparkDriverInput = rdd.collect();
+                resultRdd = ImmutableList.copyOf(prestoSparkTaskExecutorFactoryProvider.get().create(
+                        0,
+                        0,
+                        serializedTaskDescriptor,
+                        new PrestoSparkTaskInputs(ImmutableMap.of(remoteSource.getId().toString(), sparkDriverInput.iterator())),
+                        taskStatsCollector));
+                commit();
+            }
+            catch (RuntimeException executionFailure) {
+                try {
+                    rollback();
+                }
+                catch (RuntimeException rollbackFailure) {
+                    if (executionFailure != rollbackFailure) {
+                        executionFailure.addSuppressed(rollbackFailure);
+                    }
+                }
+                queryCompletedEvent(Optional.of(executionFailure));
+                throw executionFailure;
+            }
+
+            // successfully finished
+            queryCompletedEvent(Optional.empty());
+
+            ConnectorSession connectorSession = session.toConnectorSession();
+            return resultRdd.stream()
+                    .map(Tuple2::_2)
+                    .map(this::deserializePage)
+                    .flatMap(page -> getPageValues(connectorSession, page, rootFragment.getTypes()).stream())
+                    .collect(toList());
+        }
+
+        public List<Type> getOutputTypes()
+        {
+            return prestoSparkPlan.getPlan().getFragment().getTypes();
+        }
+
+        public Optional<String> getUpdateType()
+        {
+            return updateType;
+        }
+
+        private Page deserializePage(SerializedPrestoSparkPage page)
+        {
+            return pagesSerde.deserialize(readSerializedPages(Slices.wrappedBuffer(page.getBytes()).getInput()).next());
+        }
+
+        private static List<List<Object>> getPageValues(ConnectorSession connectorSession, Page page, List<Type> types)
```

Comment:
nit: Add a TODO about potential column re-ordering for `OutputNode`. Or just fix it in this PR if it's straightforward (should be if you can get the mapping right? :)

Reply:
Hmm, I'm just looking how is the `enforceLayoutProcessor` added: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L2494.

Since it is always added in the LocalExecutionPlanner unconditionally, it might be a different issue that i thought it is. So it might take some time to debug and figure out what is going on. Let's put it on our laundry list and address it with a separate PR.